src/main/java/frc/robot/commands/Shoot.java

```diff
@@ -28,26 +28,56 @@
 
 // TODO make the version with delay have correct math
 public class Shoot extends Command {
+        // the subsystems
         private final Shooter shooter;
-        public final Arm arm;
-        public final Drivetrain drive;
+        private final Arm arm;
+        private final Drivetrain drive;
         private final StorageIndex index;
 
+        // timers prevent faster-than-real-time simulation....
         private final Timer shootTimer = new Timer();
 
-        // for testing sakes
+        /** Speaker pose determined when the Shoot command runs initialize() */
+        public Pose3d speakerPose;
+
+        // These are the AprilTags on the speakers. Used to restrict search.
+        private static final int[] aprilTagsSpeaker = {3, 4, 7, 8};
+        // The empty set to disable useOnly.
+        private static final int[] aprilTagsNull = {};
+
+        // public for testing sakes; package frc.robot.util looks at these values; those tests should be elsewhere...
         public double horiz_angle;
         public double vert_angle;
         public double exit_vel;
         public Pose3d displacement;
         public double v_rx;
         public double v_ry;
         // TODO: put in constants for other commands to use
-        private final double REST_VEL = 0; // TODO: determine the fastest idle note-exit velocity that won't kill the
-                                           // battery.
+        
+        private final double REST_VEL = 0;
+        // TODO: determine the fastest idle note-exit velocity that won't kill the battery.
+
+        /**
+         * Shooter height when the arm is at the standbySetpoint.
+         * <p>
+         * This value should be in the Shooter subsystem.
+         * <p>
+         * The arm height is the pivot height + armLength * sin(standbyAngle).
+         * This calculation seems wrong because the shooter sits on top of the arm.
+         * The shooter exit is not level with the pivot when the arm is level.
+         */
         public static final double shooterHeight = ArmConstants.ARM_LENGTH*Math.sin(ArmConstants.standbySetpoint) + ArmConstants.PIVOT_HEIGHT;
+
+        /**
+         * Shooter position relative to the robot center.
+         * <p>
+         * This value should be in the Shooter subsystem.
+         * <p>
+         * The shooter exit is the pivot x + armLength * cos(standbyAngle).
+         */
         public static final double shooterOffset = ArmConstants.PIVOT_X + ArmConstants.ARM_LENGTH * Math.cos(ArmConstants.standbySetpoint);
 
+        // TODO: why should this command know anything about about tags?
```

Comment:
We have it disabled in another branch, but this will make sure it can see the speaker before shooting. If odometry is wrong enough that it doesn't (points to some other point), we shouldn't shoot.